#411 Data processing and extraction nodes and script docs (Houdini) (#413)
Inbibo/adonisfx_docs: docs/houdini/tools/data_extraction_tool.md

## Requirements

To use the AdnMLDataExtraction TOP HDA, set the following parameters:

- **Rest Skin Op Path**: SOP path to the rest skin geometry.
- **Sim Skin Op Path**: SOP path to the processed simulated skin. This should usually point to the first output of the [AdnMLDataProcessing HDA](../tools/ml_data_processing), commonly named **ADN_OUT_SKIN**.
- **Joints Op Path**: SOP path to the selected ML joints. This should usually point to the second output of the [AdnMLDataProcessing HDA](../tools/ml_data_processing), commonly named **ADN_OUT_ML_JOINTS**.
- **Joint Names List**: List of joints to extract. This list is populated from the **adnMLJoints** point group found in the joints SOP.
- **Export Folder Path**: Destination folder where the extracted data will be written.

The following optional parameter can also be provided:

- **Muscle Op Path**: SOP path to geometries affected by an **AdnMuscle** or **AdnRibbonMuscle** solver. This is required only when extracting muscle activation data.

> [!NOTE]

> > The **Joints Op Path** should point to the **ADN_OUT_ML_JOINTS** output generated by the [AdnMLDataProcessing HDA](../tools/ml_data_processing). This output is expected to contain the selected ML joints in the **adnMLJoints** point group. Pressing **Set Joint List From Op** reads this group and populates the **Joint Names List** parameter.

Reviewer: I would also add here that rest skin and sim skin must have same topology.

Author: Done

    Pressing this button finds the **adnMLJoints** point group from the geometry provided in **Joints Op Path**, then populates and formats the entries in **Joint Names List**.

    The joints listed in **Joint Names List** are the joints that will be used as input data for the ML extraction process.

<figure style="width:90%; margin-left:5%" markdown>
  ![AdnMLDataExtraction TOP HDA joint selection](../images/data_extraction_joint_selection.png)
  <figcaption><b>Figure 2</b>: Joint list populated from the <b>adnMLJoints</b> point group.</figcaption>
</figure>

7. Optionally enable **Extract Muscle Activation Data**.

    Enable **Extract Muscle Activation Data** if muscle activation values should be included in the extracted dataset.

    When this option is enabled, set **Muscle Op Path** to the SOP path of the muscle geometry. This can point to a merge node containing all muscles affected by **AdnMuscle** or **AdnRibbonMuscle** solvers, or to individual muscle output nodes using the **ADN_OUT_** prefix.

    The data extraction process will read the activation values of the underlying muscles and include them in the learning data.

8. Configure the frame settings.

>     Use the **Frame Settings** parameters to define which frames should be recorded during data extraction.

Reviewer: If I'm not mistaken we usually refer to attributes/parameters in italic style.

Author: Done

> # AdnMLDataExtraction TOP HDA

Author: Add node screenshot

> 

Reviewer 2: Add "NOTE" box to notify that this feature is available only for ML bundle.

Author: Done

> The **AdnMLDataExtraction TOP HDA** is used to extract machine learning training data from a processed Adonis simulation setup.

Reviewer 2: As you know this can also be used to extract non-Adonis data.

```suggestion
The **AdnMLDataExtraction TOP HDA** is used to extract data for training Adoinis ML models.
```

Author: Done

This HDA is used after the [AdnMLDataProcessing HDA](../tools/ml_data_processing) has prepared the simulation skin and selected ML joints. The AdnMLDataExtraction TOP HDA reads the processed skin, the rest skin, and the selected joint data, then exports the data required for the machine learning workflow.

The joint list used for extraction is expected to come from the processed ML joint output of the [AdnMLDataProcessing HDA](../tools/ml_data_processing), usually exposed through a **Null** node named **OUT_JOINTS**. This output contains a point group named *adnMLJoints*, which defines the joints to include in the extracted input data.

> Optionally, the AdnMLDataExtraction TOP HDA can also extract muscle activation data. When enabled, this allows the extraction process to include activation values from the underlying muscles for the learning process.

Reviewer 2: ```suggestion
Optionally, the AdnMLDataExtraction TOP HDA can also extract muscle activation data. When enabled, this allows the extraction process to include dynamic material properties prediction for AdnSmartTissue support of the trained ML models.
```

Author: Done

| *Set Joint List From Op* | Button |  | Finds the *adnMLJoints* group from the geometry specified in *Joints Op Path*, then populates and formats the entries in *Joint Names List*. |
| *Joint Names List* | String List |  | List of joint names to extract as ML input data. This list is usually populated using *Set Joint List From Op*. |

### Muscles

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| *Extract Muscle Activation Data* | Boolean | False | Enables extraction of muscle activation data. When enabled, the extraction process reads activation values from the muscles provided in *Muscle Op Path*. |
| *Muscle Op Path* | SOP Path |  | SOP path to geometries that have an **AdnMuscle** or **AdnRibbonMuscle** solver affecting them. This can be a single operator path pointing to a merge node containing all muscles, or individual operator paths prefixed with **ADN_OUT_**. |

### Frame Settings

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |

> | *Record Frame Windows* | Boolean | False | Enables the use of specific frame windows for the data extraction process. Frame windows must be defined as separate frame ranges, for example: `[[1, 5], [220, 600], [1023, 1500]]`. |

Reviewer 2: Missing frame windows entry. Please split the description as this one is tagged as boolean.

Author: Done

### Output

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| *Export Folder Path* | Folder Path |  | Destination folder where the extracted data will be written. |
| *Force Overwrite* | Boolean | False | Allows existing extraction data in the output folder to be overwritten. |

## Result

After cooking the AdnMLDataExtraction TOP HDA, the extraction process writes the machine learning data to the folder specified in *Export Folder Path*.

The extracted data is generated from:

- The rest skin specified in *Rest Skin Op Path*.
- The processed simulated skin specified in *Sim Skin Op Path*.
- The selected ML joints specified in *Joints Op Path* and listed in *Joint Names List*.

> - The optional muscle activation data specified through *Muscle Op Path*, if *Extract Muscle Activation Data* is enabled.

Reviewer 3: Do we need to expose again the inputs here? I think we are reiterating here, we could go directly with "The following files are generated ..."

Author: Done

The rest skin and processed simulated skin must have the same topology so the extracted output data can be computed from matching points.

The following files are generated in the target export folder:

- `inputs.csv`: Contains the extracted input data, including the selected joint transforms.
- `outputs.csv`: Contains the extracted output data, including the simulated skin displacement data and optional data required for AdnSmartTissue material properties prediction.
- `joints.json`: Contains the exported joint hierarchy information used to associate the joint transform data with the extracted joints.
- `extraction_config.json`: Contains the configuration used for the extraction process, including the input paths, joint list, frame settings, and export settings.

> [!NOTE]
> If the extraction is interrupted or fails before completion, some of these files may already exist in the export folder but contain partial data.

## Limitations

> - The AdnMLDataExtraction TOP HDA requires data prepared by the [AdnMLDataProcessing HDA](../tools/ml_data_processing).

Reviewer 3: I don't know if we need this list of limitations. To me, they are again the requirements, not limitations per se, like something that it is not supported yet and will be eventually supported in future versions. You decide, but maybe we can just remove this section

Author: Done
